### ISSUE TITLE
Add option to apply torch.softmax or disable torch.log internally

### ISSUE THREAD

**Issue by ToddMorrill** [skorch-dev/skorch#637]
From what I've gathered, there are basically 3 options for losses for classification tasks when using Skorch.

1)  Have your network output probabilities from a `torch.softmax` layer and then use `nn.NLLLoss`, where Skorch is applying `orch.log` internally. But this has been shown to be numerically unstable. [Overview](https://discuss.pytorch.org/t/why-there-is-no-log-operator-in-implementation-of-torch-nn-nllloss/16610), [Details](https://timvieira.github.io/blog/post/2014/02/11/exp-normalize-trick/)
- No solution: based on my understanding of the links about, I would advise against doing this.

2) Have your network output from `nn.logsoftmax` and then use `nn.NLLLoss`, which is numerically stable. Currently Skorch is applying `torch.log` to the output of the network when `nn.NLLLoss` is specified. So `torch.log` is applied twice and the model doesn't train.
- Solution: have an option to disable the application of `torch.log` internally before `nn.NLLLoss`.

3) Have your network output raw logits (i.e. don't apply any sort of softmax) and then use `nn.CrossEntropyLoss`, which is numerically stable. Skorch will handle training but then `model.predict_proba` breaks down because you're getting raw logits. Obviously I can just apply the softmax function but it breaks the semantics of the sklearn API.
- Solution: have an option to apply of `torch.softmax` internally before `model.predict_proba`.

**Comment by BenjaminBossan** [comment]
Hi Todd,

Your assessment of 1 is correct, though just from my personal experience, I haven't found it ever to matter (maybe I was just lucky).

Your assessment of 2 is a bit incomplete. If the module returns log softmax, the result from `predict_proba` wouldn't be probabilities, so you'd need to change that as well.

Your assessment of 3 is correct.

Regarding solutions, instead of having complicated options of what transformation to apply when, we wanted to go with a solution to make it easy to implement your own transformation (leaving the defaults as they are). [This is the PR](https://github.com/skorch-dev/skorch/pull/580). Unfortunately, it's been asleep for some time, hopefully it will be finished eventually.

**Comment by ToddMorrill** [comment]
Alright, I'll stay tuned. Thanks for the update.
